src/Transport/Sending/CustomizeNativeMessageExtensions.cs

```diff
@@ -0,0 +1,63 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using Microsoft.Azure.ServiceBus;
+    using NServiceBus.Extensibility;
+
+    /// <summary>
+    /// Allows the users to customize outgoing native messages.
+    /// </summary>
+    /// <remarks>
+    /// The behavior of this class is exposed via extension methods.
+    /// </remarks>
+    public static class CustomizeNativeMessageExtensions
+    {
+        internal const string CustomizationHeader = "$ASB.CustomizationId";
+
+        /// <summary>
+        /// Allows customization of the outgoing native message sent using <see cref="IMessageSession"/>.
+        /// </summary>
+        /// <param name="options">Option being extended.</param>
+        /// <param name="customization">Customization action.</param>
+        public static void CustomizeNativeMessage(this ExtendableOptions options, Action<Message> customization)
+        {
+            if (options.GetHeaders().ContainsKey(CustomizationHeader))
+            {
+                throw new InvalidOperationException("Native outgoing message has already been customized. Do not apply native outgoing message customization more than once per message.");
+            }
+
+            var customizationId = Guid.NewGuid().ToString();
+            options.SetHeader(CustomizationHeader, customizationId);
+
+            var nativePropertiesCustomizer = options.GetExtensions().GetOrCreate<NativeMessageCustomizer>();
+            if (!nativePropertiesCustomizer.Customizations.TryAdd(customizationId, customization))
+            {
+                throw new Exception("Failed to apply an outgoing message customization");
+            }
+        }
+
+        /// <summary>
+        /// Allows customization of the outgoing native message sent using <see cref="IMessageHandlerContext"/>.
+        /// </summary>
+        /// <param name="options">Option being extended.</param>
+        /// <param name="context"><see cref="IMessageHandlerContext"/> used to dispatch messages in the message handler.</param>
+        /// <param name="customization">Customization action.</param>
+        public static void CustomizeNativeMessage(this ExtendableOptions options, IMessageHandlerContext context, Action<Message> customization)
```

Comment:
If we'd use IPipelineContext here I think we could enable this in the whole receive pipeline including behaviours and not just limit it to handlers. Thoughts?

Reply:
I can't see a reason why not to open up where the customization can be done.
Saying that the evidence I have so far indicates it would be done in the context of the incoming message. Which is not available on `IPipelineContext`. The earliest that has _anything_ related to the incoming message would be `IMessageProcessingContext` with the incoming message headers.

Reply:
I don't think that is true. All send operations and extensions on the `IPipelineContext` have options and the only requirement we have is that we can attach a header that we later remove and have access to the extension bag. This is all given on pipeline context. Not sure I understand why incoming headers are relevant

Reply:
The outgoing message customization is there to allow manipulation of the native dispatched message in some business context. That context is the incoming message with the information either in the body or the headers. The [original ask](https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/issues/202#issuecomment-724921720) for this feature was driven by this requirement. `IPipelineContext` gives you none of those two. So what good it will do other than being able to assign some arbitrary values to the outgoing native message? Hence my comment about `IMessageProcessingContext` which at least has the incoming headers. Don't focus on the mechanical ability to modify the outgoing message, rather _why_ and _what_ that customization is needed to begin with.

Reply:
I do agree in general with what you are saying but then you limit the feature to the original use case. What if the user has other needs and wants to customize the native header with other contextual information as part of a behavior pipeline? Then you would actively prevent such use case scenarios to the pipeline stages that expose `IMessageProcessingContext`. I'll send in a spike to show what I mean

Reply:
Raised as a draft https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/pull/205

What the spike confirmed is that since we only care about the context because we need to customize things on the incoming path but we are not exposing it back into the lambda anything concrete that would need to be accessed from the context needs to be extracted before the action delegate anyway and then passed into the action delegate as closure. So even accessing incoming headers would be

```
                    var someHeader = context.MessageHeaders["SomeHeader"];
                    var publishOptions = new PublishOptions();
                    publishOptions.CustomizeNativeMessage(context, m => m.Label = someHeader);
```

This confirms my conclusion that there is no reason to really restrict it to a specific context.

Another thought that came up is if it would make sense to have the context overload do something like

```
                    var publishOptions = new PublishOptions();
                    publishOptions.CustomizeNativeMessage(context, (ctx, m) => m.Label = ctx.MessageHeaders["SomeHeader"]);
```

the upside would be that the contextual information can be accessed in the lambda and it clearly differentiates it, even more, from the `CustomizeNativeMessage` that is "context-free". The drawback would be that we need to capture the context and make them live longer than intended which could cause problems down the line since they are not designed to outlive the stages they are supposed to be used in.

Reply:
> publishOptions.CustomizeNativeMessage(context, (ctx, m) => m.Label = ctx.MessageHeaders["SomeHeader"]);

Feels very awkward. Combined with the fact that you’ve outlined being problematic, I don’t think we should consider it.

> What if the user has other needs and wants to customize the native header with other contextual information as part of a behavior pipeline?

This is what I’m trying to avoid, “what-if”s. Building something concrete and changing when there’s a need rather building based on a guess is what I’m in favour of. Saying that, your original comment did not explain what you had in mind. Rather than “use IPipelineContext” the #205 PR is _adding_ an option of using `IPipelineContext` as an additional overload. That would be a good addition.

Reply:
The comment about "use pipeline context" was on the method

```
public static void CustomizeNativeMessage(this ExtendableOptions options, IMessageHandlerContext context, Action<Message> customization)
```

I apologize if that was not precise enough. I felt our communication was going sideways because of a misunderstanding the I probably caused. That's why I raised the PR

Reply:
Yeah, wasn't clear but we've sorted it out and 👍 now.